katsdpimager/report.py

```diff
@@ -0,0 +1,356 @@
+
+    def make_plot_flux(self, source: bokeh.models.ColumnDataSource) -> bokeh.model.Model:
+        fig = bokeh.plotting.figure(
+            x_axis_label=f'Frequency ({FREQUENCY_PLOT_UNIT})',
+            y_axis_label=f'Flux ({FLUX_PLOT_UNIT})',
+            x_range=self.frequency_range,
+            y_axis_type='log',
+            sizing_mode='stretch_width',
+        )
+        si_format = bokeh.models.CustomJSHover(code="""
+            const threshold = [1e-9, 1e-6, 1e-3, 1e0, 1e3,  1e6,  1e9, 1e12];
+            const scale =     [1e12,  1e9,  1e6, 1e3, 1e0, 1e-3, 1e-6, 1e-9, 1e-12];
+            const suffix =    [ "p",  "n",  "µ", "m",  "",  "k",  "M",  "G", "T"];
+            var i = 0;
+            while (i < threshold.length && value >= threshold[i])
+                i++;
+            return sprintf(format, value * scale[i]) + " " + suffix[i];
+            """)
+        fig.add_tools(bokeh.models.HoverTool(
+            tooltips=[
+                ('Frequency', f'@frequency{{0.0000}} {FREQUENCY_PLOT_UNIT}'),
+                ('Channel', '$index'),
+                ('Peak', f'@peak{{%.3f}}{FLUX_PLOT_UNIT}'),
+                ('Noise', f'@noise{{%.3f}}{FLUX_PLOT_UNIT}'),
+                ('Predicted noise', f'@predicted_noise{{%.3f}}{FLUX_PLOT_UNIT}')
+            ],
+            formatters={
+                '@peak': si_format,
+                '@noise': si_format,
+                '@predicted_noise': si_format
+            }
+        ))
+        fig.line(x='frequency', y='peak', source=source, name='peak',
+                 line_color=PALETTE[0], legend_label='Peak')
+        fig.line(x='frequency', y='noise', source=source, name='noise',
+                 line_color=PALETTE[1], legend_label='Noise')
+        if self.predicted_natural_noise is not None:
+            fig.line(x='frequency', y='predicted_noise', source=source, name='predicted_noise',
+                     line_color=PALETTE[2], legend_label='Predicted noise')
+        self._add_channel_range(fig)
+        return fig
+
+    def make_plot_elevation(self, source: bokeh.models.ColumnDataSource) -> bokeh.model.Model:
+        unit_str = ANGLE_PLOT_UNIT.to_string('unicode')  # deg -> °
+        fig = bokeh.plotting.figure(
+            x_axis_label='Time (UTC)', x_axis_type='datetime',
+            y_axis_label=f'Elevation ({unit_str})',
+            x_range=self.time_range,
+            y_range=[0.0, (90 * u.deg).to_value(ANGLE_PLOT_UNIT)],
+            sizing_mode='stretch_width'
+        )
+        fig.add_tools(bokeh.models.HoverTool(
+            tooltips=[
+                ('Time', '@time{%Y-%m-%d %H:%M:%S}'),
+                ('Elevation', f'@elevation {unit_str}')
+            ],
+            formatters={
+                '@time': 'datetime'
+            }
+        ))
+        fig.circle(x='time', y='elevation', source=source, line_color=PALETTE[0])
+        # TODO: plot LST on the opposite axis?
+        return fig
+
+    def make_plot_parallactic_angle(
```

Comment:
While parallactic angle is useful. A more natural thing to plot that astronomers tend to understand better in terms of what they are getting out of the observation is hour angle. Could you plot both parallactic angle and hour angle in this plot using different coloured lines?

Reply:
I plotted PA because that determines how much of the beam sources move through. Interface-wise that wouldn't be a problem - does katdal provide hour angle?

Reply:
Not sure, but you should be able to get LST from a timestamp.

Reply:
I was suggesting hour angle because it gives an indication of the (u,v) coverage on the target. Actually - perhaps just plotting the (u,v) coverage might be better - and will work quite nicely with bokeh's zoom functionality.

Reply:
I see HA is just an offset to sidereal time - so would a plot of HA really give much more information than knowing the time spent on target (I guess together with how long the gaps are for calibrator visits etc - but that should come out on the elevation plot)?

I'll think about UV coverage, but there might be some complications:
- It can't take into account flagging - at the moment that means it'll be a gross misrepresentation for some channels until I get around to skipping imaging all together for channels in the RFI mask.
- It might be too much data. There will be thousands of tracks, and presumably they need to be sampled in time as well. I don't want the report file to end up as hundreds of megabytes. Maybe I can figure out equations for the ellipses traced and just plot elliptic arcs, but it's more than I want to try to squeeze into this PR.

Reply:
>I see HA is just an offset to sidereal time - so would a plot of HA really give much more information than knowing the time spent on target (I guess together with how long the gaps are for calibrator visits etc - but that should come out on the elevation plot)?

If the target is tracked only a few times over a long period of time (with other targets in the same observation), plotting the hour angle gives an indication of not only the time spent on target, but the expected 'quality' of your PSF (wider hour angle coverage = 'better' PSF to first order). The elevation plot will sort of show you this as well, its just more traditionally done by plotting hour angle. In reality there aren't many MeerKAT observation like what I am describing I know of (yet!) so don't worry about it.

>It might be too much data. There will be thousands of tracks, and presumably they need to be sampled in time as well. I don't want the report file to end up as hundreds of megabytes. Maybe I can figure out equations for the ellipses traced and just plot elliptic arcs, but it's more than I want to try to squeeze into this PR.

I've tended to do it in the past by sub-sampling each baseline by a few points in time - you are right that it will be too much data if you plot it all. Don't worry about it for now it certainly isn't critical either (just a suggestion).